### PR TITLE
fix: resolve conflict descendants

### PR DESCRIPTION
**Pull request by zhangsoledad**
### What problem does this PR solve?

Previously, pending transaction removed but descendants remain.

### What is changed and how it works?

1. remove conflict and descendants.
2. transaction expiration.
3. non-mine mode.

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code ci-runs-only: [ quick_checks,linters ]

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

File: tx-pool/src/process.rs [hunk 1]
```diff
@@ -1380,81 +1379,77 @@ fn _update_tx_pool_for_reorg(
...
+    tx_pool.remove_committed_txs(attached.iter(), callbacks, detached_headers);
+    tx_pool.remove_by_detached_proposal(detached_proposal_id.iter());
 
     let mut entries = Vec::new();
     let mut gaps = Vec::new();
 
+    // mine mode:
     // pending ---> gap ----> proposed
     // try move gap to proposed
+    if mine_mode {
```
Comment (quake) [inline tx-pool/src/process.rs:1400]: line 1394 and 1395 should be moved to this conditional block:
```
    let mut entries = Vec::new();
    let mut gaps = Vec::new();
```

File: tx-pool/src/process.rs [hunk 2]
```diff
@@ -1380,81 +1379,77 @@ fn _update_tx_pool_for_reorg(
...
-    for (cycles, entry) in entries {
-        let tx_hash = entry.transaction().hash();
-        if let Err(e) = tx_pool.proposed_rtx(cycles, entry.size, entry.rtx.clone()) {
-            debug!("Failed to add proposed tx {}, reason: {}", tx_hash, e);
-            callbacks.call_reject(tx_pool, &entry, e.clone());
-        } else {
-            callbacks.call_proposed(tx_pool, &entry, false);
+        for (cycles, entry) in entries {
+            let tx_hash = entry.transaction().hash();
+            if let Err(e) = tx_pool.proposed_rtx(cycles, entry.size, entry.rtx.clone()) {
```
Comment (quake) [inline tx-pool/src/process.rs:1433]: the parameter `cache_entry: Option<CacheEntry>` of `proposed_rtx`  and `gap_rtx` fn may be changed to `cache_entry: CacheEntry`

File: tx-pool/src/process.rs [hunk 3]
```diff
@@ -1380,81 +1379,77 @@ fn _update_tx_pool_for_reorg(
...
     let mut entries = Vec::new();
     let mut gaps = Vec::new();
 
+    // mine mode:
     // pending ---> gap ----> proposed
     // try move gap to proposed
+    if mine_mode {
+        tx_pool.gap.remove_entries_by_filter(|id, tx_entry| {
+            if snapshot.proposals().contains_proposed(id) {
+                entries.push((
```
Comment (quake) [inline tx-pool/src/process.rs:1403]: suggest to change the code:
```
   entries.push(tx_entry.clone();
```

and build the `CachedEntry` in the next loop:
```
for entry in entries {
  tx_pool.proposed_rtx(CacheEntry::completed(tx_entry.cycles, tx_entry.fee), entry.size, entry.rtx.clone())
}
```

File: tx-pool/src/component/pending.rs [hunk 4]
```diff
@@ -138,54 +164,71 @@ impl PendingQueue {
...
-            return Some(entry);
         }
-        None
+        removed
     }
 
-    pub(crate) fn remove_entry(&mut self, id: &ProposalShortId) -> Option<TxEntry> {
-        let removed = self.inner.remove(id);
+    pub(crate) fn get_descendants(&self, entry: &TxEntry) -> Vec<ProposalShortId> {
+        use std::collections::VecDeque;
```
Comment (quake) [inline tx-pool/src/component/pending.rs:193]: for consistent code style, please move to the header.

File: tx-pool/src/component/pending.rs [hunk 5]
```diff
@@ -138,54 +164,71 @@ impl PendingQueue {
...
 
-    pub(crate) fn remove_committed_tx(
-        &mut self,
-        tx: &TransactionView,
-        related_out_points: &[OutPoint],
-    ) -> Option<TxEntry> {
-        let inputs = tx.input_pts_iter();
-        let id = tx.proposal_short_id();
+    pub(crate) fn remove_entry(&mut self, id: &ProposalShortId) -> Option<TxEntry> {
+        let removed = self.inner.remove(id);
```
Comment (quake) [inline tx-pool/src/component/pending.rs:168]:
```suggestion
        self.inner.remove(id).map(|entry| {
            self.remove_entry_relation(&entry);
            entry
        })
```

File: tx-pool/src/component/pending.rs [hunk 6]
```diff
@@ -138,54 +164,71 @@ impl PendingQueue {
...
+            let mut empty = false;
+
+            if let Some(ids) = self.inputs.get_mut(&i) {
+                ids.remove(&tx_short_id);
+                empty = ids.is_empty();
+            }
+
+            if empty {
+                self.inputs.remove(&i);
+            }
```
Comment (quake) [inline tx-pool/src/component/pending.rs:231]: suggest to use `Entry::Occupied`, which can reduce one hash lookup, this suggestion also apply to `// remove dep` code block

```suggestion
            if let Entry::Occupied(mut entry) = self.inputs.entry(i) {
                let empty = {
                    let ids = entry.get_mut();
                    ids.remove(&tx_short_id);
                    ids.is_empty()
                };
                if empty {
                    entry.remove();
                }
            }
```

**Review by quake — CHANGES_REQUESTED** [review]

File: tx-pool/src/process.rs [hunk 7]
```diff
@@ -1380,81 +1379,72 @@ fn _update_tx_pool_for_reorg(
...
-            gaps.push((
-                Some(CacheEntry::completed(tx_entry.cycles, tx_entry.fee)),
-                tx_entry.clone(),
-            ));
-            true
-        } else {
-            false
-        }
-    });
+        // Some(CacheEntry::completed(tx_entry.cycles, tx_entry.fee)),
```
Comment (quake) [inline tx-pool/src/process.rs:1401]: please remove this line

**Review by quake — APPROVED** [review]

**Review by yangby-cryptape — APPROVED** [review]

**Comment by quake** [comment]
bors r=quake,yangby-cryptape

**Pull request merged**